% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/runContrasts.R
\name{runContrasts}
\alias{runContrasts}
\title{Function  runContrasts}
\usage{
runContrasts(
  dgeObj,
  designMatrixName,
  contrastList,
  contrastSetName = fitName,
  runTopTable = TRUE,
  runTopTreat = FALSE,
  FoldChangeThreshold = 1.5,
  PvalueThreshold = 0.01,
  FDRthreshold = 0.1,
  runEBayes = TRUE,
  robust = TRUE,
  proportion = 0.01,
  Qvalue = FALSE,
  IHW = FALSE,
  verbose = FALSE
)
}
\arguments{
\item{dgeObj}{A DGEobj object containing a Fit object and design matrix (required)}

\item{designMatrixName}{The name of the design matrix within dgeObj to use for
contrast analysis  (required)}

\item{contrastList}{A named list of contrasts  (required)}

\item{contrastSetName}{Name for the set of contrasts specified in contrastList.  Defaults
to "fitName_cf".  Change only if you need to create 2 or more contrast sets from the same fit.}

\item{runTopTable}{runs topTable on the specified contrasts (Default = TRUE)}

\item{runTopTreat}{runs topTreat on the specified contrasts (Default = FALSE)}

\item{FoldChangeThreshold}{Only applies to TopTreat (Default = 1.5)}

\item{PvalueThreshold}{Default = 0.01}

\item{FDRthreshold}{Default = 0.1}

\item{runEBayes}{Runs eBayes after contrast.fit (default = TRUE)}

\item{robust}{eBayes robust option (default = TRUE)}

\item{proportion}{Proportion of genes expected to be differentially expressed
(used by eBayes) (Default = 0.01)}

\item{Qvalue}{Set TRUE to include Qvalues in topTable output (Default=FALSE)}

\item{IHW}{Set TRUE to add FDR values from the IHW package (Defulat=FALSE)}

\item{verbose}{Set TRUE to print some information during processing (Default=FALSE)}
}
\value{
The DGEobj with contrast fits and topTable/topTreat dataframes added.
}
\description{
Take a DGEObj and a named list of contrasts to build. The DGEobj must
contain a limma Fit object and associated DesignMatrix. Returns the DGEobj with
contrast fit(s), contrast matrix and topTable/topTreat dataframes added.
}
\details{
The contrastList is a named list composed of column names from the DesignMatrix
of the DGEobj.  Each contrast is named to give
it a short recognizable name.

Example contrastList \cr

contrastList = list( \cr
   T1 = "treatment1 - control", \cr
   T2 = "treatment2 - control" \cr
) \cr

where treatment1, treatment2 and control are columns in the DesignMatrix

The returned ContrastAnalysis list contains the following objects:
\itemize{
   \item{"ContrastMatrix"} {a matrix}
   \item{"Fit.Contrasts"} {a Fit object}
   \item{"TopTableList"} {a List of dataframes}
   \item{"TopTreatList"} {a List of dataframes}
}
}
\examples{
#run defaults
myDgeObj  = runContrasts (myDgeObj, myFitName, ConstrastList)
myDgeObj  = runContrasts (myDgeObj, myFitName, ConstrastList, runTopTable = TRUE
       runTopTreat=TRUE, FoldChangeThreshold = 1.25)

}
\author{
John Thompson, \email{jrt@thompsonclan.org}
}
\keyword{DGE;}
\keyword{DGEobj}
\keyword{RNA-Seq;}
\keyword{contrasts;}
